src/cart/crt_hg.c

```diff
@@ -327,6 +467,88 @@ crt_hg_addr_lookup(struct crt_hg_context *hg_ctx, const char *name,
 	return rc;
 }
 
+int
+crt_hg_remove_addr(hg_class_t *hg_class, hg_addr_t hg_addr)
+{
+	hg_return_t     ret = HG_SUCCESS;
+
+	D_DEBUG(DB_TRACE, "removing hg addr %" PRIu64 "\n",
+			(uint64_t)hg_addr);
+	D_DEBUG(DB_TRACE, "removing  hg class %" PRIu64 "\n",
+			(uint64_t)hg_class);
+	ret = HG_Addr_set_remove(hg_class, hg_addr);
+	if (ret != HG_SUCCESS) {
+		D_ERROR("HG_Addr_set_remove) failed, hg_ret %d.\n", ret);
+		D_GOTO(out, ret = -DER_HG);
+	}
+	ret = HG_Addr_free(hg_class, hg_addr);
+	if (ret != HG_SUCCESS) {
+		D_ERROR("HG_Addr_free() failed, hg_ret %d.\n", ret);
+		D_GOTO(out, ret = -DER_HG);
+	}
+
+out:
+	return ret;
+}
+
+int
+crt_hg_remove_client_id(uint64_t client_id)
+{
+	int				ret = 0;
+	d_list_t			*halink;
+	struct crt_ha_mapping		*ha;
+	struct ha_entry			ha_value;
+	struct crt_ha_list_entry	*entry;
+	struct crt_context		*crt_ctx;
+
+	if (crt_gdata.cg_na_plugin != CRT_NA_OFI_PSM2)
+		D_GOTO(out, ret);
+
+	if (!crt_is_service())
+		D_GOTO(out, ret);
+
+	D_RWLOCK_RDLOCK(&crt_gdata.cg_rwlock);
+
+	d_list_for_each_entry(crt_ctx, &crt_gdata.cg_ctx_list, cc_link) {
+		/*Can this lock create a deadlock? TBD */
+		D_RWLOCK_WRLOCK(&crt_ctx->cc_ha_hash_table_rwlock);
+		halink = d_hash_rec_find(&crt_ctx->cc_ha_hash_table,
+			(void *)&client_id, sizeof(client_id));
+		if (!halink) {
+			D_ERROR("client=%" PRIu64 " not in the hash table\n",
+					client_id);
```

Comment:
we should return error at this point to the caller (make sure to unlock taken locks first)